leader/leader.go

```diff
@@ -124,229 +129,218 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
 }
 
 // AddWatch starts watching the key for changes and sending them
-// to the valuesC, the watch is stopped
-func (l *Client) AddWatch(key string, retry time.Duration, valuesC chan string) {
-	prefix := fmt.Sprintf("AddWatch(key=%v)", key)
+// to the valuesC.
+// The watch can be stopped either with the the specified context or Client.Close
+func (l *Client) AddWatch(ctx context.Context, key string, retry time.Duration, valuesC chan string) {
 	api := client.NewKeysAPI(l.client)
-
-	go func() {
+	renew := func() (client.Watcher, *client.Response) {
+		b := NewUnlimitedExponentialBackOff()
 		var watcher client.Watcher
 		var resp *client.Response
-		var err error
-
-		ctx, closer := context.WithCancel(context.WithValue(context.Background(), "prefix", prefix))
-		go func() {
-			<-l.closeC
-			closer()
-		}()
-
-		backoff := &FreebieExponentialBackOff{
-			InitialInterval:     500 * time.Millisecond,
-			RandomizationFactor: 0.5,
-			Multiplier:          1.5,
-			MaxInterval:         10 * time.Second,
-			MaxElapsedTime:      30 * time.Second,
+		backoff.Retry(func() (err error) {
+			watcher, resp, err = l.recreateWatchAtLatestIndex(ctx, api, key, retry)
+			return trace.Wrap(err)
+		}, b)
+		return watcher, resp
```

Comment:
Yes, it omits the error on purpose as the retry loop is not time-limited and the `watcher == nil` is special in the sense that it indicates the event of closing the client.
The whole retry is only supposed to be cancelled (via a context or calling `client.Close`) - this is all due to the way API is structured (e.g. all errors are handled internally by the `leader` package).

The alternative is moving the error handling out of the leader loops into the client (analogue to https://github.com/docker/leadership, which has a much simpler implementation due to this).